### ISSUE TITLE
Unable to query for MongoDB ObjectIDs

### ISSUE THREAD

**Issue by alexmcmillan** [graphql/graphql-js#1518]
I recently updated from `0.13.2` to `14.0.2` which includes breaking changes.

This introduced errors with existing queries which include MongoDB Object Ids (probably from #1382):

> ID cannot represent value: { _bsontype: \"ObjectID\", id: <Buffer 5b 96 3d bf 98 0a 04 09 85 c6 6e a1> }

Repository with complete, minimal repeatable example [here](https://github.com/alexmcmillan/graphql-mongodb-query-fail-demo):

```
const Thing = mongoose.model('Thing', new mongoose.Schema({
  id: mongoose.Schema.Types.ObjectId,
  name: String
}));

const ThingType = new GraphQLObjectType({
  name: 'thing',
  fields: function () {
    return {
      id: { type: GraphQLID },
      name: { type: GraphQLString }
    }
  }
});

const RootMutation = new GraphQLObjectType({
  name: 'CreateMutation',
  fields: {
    create: {
      type: ThingType,
      description: 'Create new thing',
      args: {
        name: {
          name: 'Name of Thing',
          type: new GraphQLNonNull(GraphQLString)
        }
      },
      resolve: (root, args) => {
        const newThing = new Thing({ name: args.name });
        newThing.id = newThing._id;
        return new Promise((res, rej) => {
          newThing.save(err => {
            if (err) return rej(err);
            res(newThing);
          });
        });
      }
    }
  }
});
```

**Comment by alexmcmillan** [comment]
Background:  I'm a learner.  Following [this tutorial](https://www.compose.com/articles/using-graphql-with-mongodb/) I encountered [this problem](https://stackoverflow.com/questions/52243415/what-is-the-correct-type-to-use-for-an-objectid-field-across-mongoose-and-graphq/52243907?noredirect=1) which was _very_ confusing trying to debug.  It would be nice if there was perhaps a clearer error message (`GraphQLID cannot represent value...` would have helped me a lot, as would an indication of how to properly cast a MongoDB ObjectID).

**Comment by sibelius** [comment]
any changes this gets into 14.0.3?

**Comment by RadAcademy** [comment]
Yes, I am still stuck using '^0.13.2' as the latest '^14.0.2' continues to give me these issues.

**Comment by GlauberF** [comment]
I would also like to know if it has been fixed for the ```^14.0.2```, 
Why am I having to use the ```^0.13.2```

**Comment by IvanGoncharov** [comment]
@RadAcademy @GlauberF We plan to include it into the next release.
However since instead of depending on `toString` we now using `toJSON` it can't be a patch release. So we are working on `14.1.0` that will include this and a few other features.
We will try to release RC ASAP.

**Comment by axe-z** [comment]
I'm a newbie at apollo, me and my aspiration .... arggh. the JS fatigue is real. Wanted to let redux go, by only using this now. here's my very small project: 

server:
```
const express = require("express");
const mongoose = require("mongoose");
require("dotenv").config({ path: "variables.env" }); //a juste besoin du path
//const bodyParser = require('body-parser');
const { ApolloServer, gql} = require('apollo-server-express');
 
const cors = require("cors"); // pour que react chiasse pas en faisant des calls

mongoose
  .connect(process.env.MONGO_URI, { useNewUrlParser: true, useCreateIndex: true })  
  .then(() => console.log("Db connecté"))
  .catch(err => console.error(err));

const app = express();

//Schema
const Blog = require('./models/Blog');
const User = require('./models/User');

// //typedefs et resolvers
const { typeDefs } = require('./schema')
const { resolvers } = require('./resolvers')
 


//pour react
const corsOption = {
  origin: "http://localhost:3000",
   credentials: true
};

app.use(cors(corsOption));


const port =  4444;

const server = new ApolloServer({
  typeDefs,
  resolvers,
  context: {
    Blog,
    User
  }
});

server.applyMiddleware({app});


app.listen({ port }, () =>
  console.log(`🚀 Server ready at http://localhost:${port}${server.graphqlPath}`),
);

```

Blog shema 
exports.typeDefs = `

```
type Blog {
  _id: ID
  title: String!
  author: String
  body: String!
  date: String
}
type Query {
  getAllBlogs: [Blog]
}
type Mutation {
  addBlog(title: String!, author: String, body: String!) : Blog
}
`;
```
resolver: 
```
exports.resolvers = {
  Query: {
    getAllBlogs: async(root, args, { Blog }) => await Blog.find().sort({
      date: -1  //-1 = desc //1 = ascendant
    })
  },
  Mutation: {
      addBlog: async(root, { title, author, body  }, { Blog }) => {
        const newBlog = await new Blog({
          title, author, body
        }).save()
        return newBlog
      }
  },
 
}
```

in react WHERE  I GET Issues with ID: 

```
import React  from 'react';
import { Query } from "react-apollo";
import gql from "graphql-tag";

const GET_ALL_BLOGS = gql`
  query {
    getAllBlogs {
      !!!!!IF I PUT _ID I GET THE ERROR !!!!!!!!!!!!!
      title
      author
      body
      date
    }
  }
`;

const App = props => {
  return (
      <div>
        <h2>My first Apollo app </h2>
        <Query query={GET_ALL_BLOGS}>
          {({ data : { getAllBlogs } , loading, error }) => {
            if (loading) return "Loading";
            if (error) return `Error!: ${error}`;
            return  (
              <ul>
                {getAllBlogs.map(el =>  <li key={el.WOULDBE_ID}>{el.title}</li>)}
              </ul>
            );
          }}
        </Query>
      </div>
  );
};


```

index has the client, but it's ok, IT's that simple right now... Not sure what I'm doing wrong

**Comment by IvanGoncharov** [comment]
@axe-z We already have a fix for this issue in #1520 
We will try to release it ASAP.

**Comment by axe-z** [comment]
oh thanks, when you are at your first steps like me , nosebleeds are ruff.

**Comment by axe-z** [comment]
since I have not a clue what's going on ... this is on the server-side only ? did install it,seems to work, after restarting .

**Comment by RadAcademy** [comment]
Your query needs to match the string type of string!  not string

**Comment by axe-z** [comment]
ok , really not sure what I'm doing wrong then
```
export const ADD_BLOG = gql`
  mutation($title: String!, $author: String!, $body: String!) {
    addBlog(title: $title, author: $author, body: $body) {
      _id
      title
      author
      body
      date
    }
  }
`;
```

**Comment by sachintbits** [comment]
I had the same issue, i resolved this issue by writing this code in the types definition file of graphql

```
import mongoose from "mongoose";

const { ObjectId } = mongoose.Types;

ObjectId.prototype.valueOf = function() {
  return this.toString();
};
```

**Comment by sibelius** [comment]
should this be done on `mongoose` instead of `graphql-js`?

@vkarpov15 does it make sense to `ObjectId` implement `valueOf`?

**Comment by vkarpov15** [comment]
@sibelius not a bad idea. Opened up a mongoose issue to track this

**Comment by sibelius** [comment]
@vkarpov15 tks, keep the good work

**Comment by infinitegachi** [comment]
still not fixed?

**Comment by vkarpov15** [comment]
I don't use graphql so I'm not certain whether this is fixed, but this might help: http://thecodebarbarian.com/whats-new-in-mongoose-54-global-schematype-configuration.html

**Comment by sibelius** [comment]
@infinitegachi try to use this mongoose PR https://github.com/Automattic/mongoose/pull/7353

**Comment by sibelius** [comment]
I've made a field resolver to help on this

```jsx
import { GraphQLNonNull, GraphQLString } from 'graphql';
import { Types } from 'mongoose';

export const mongooseIDResolver = {
  _id: {
    type: GraphQLNonNull(GraphQLString),
    description: 'mongoose _id',
    resolve: ({ _id }: { _id: Types.ObjectId}) => ({ _id: _id.toString() }),
  },
};
```

so if you want to expose an _id in your GraphQL, you need a custom type or transform ObjectId to string

**Comment by sibelius** [comment]
Types.ObjectId is for typechecking only I guess

https://github.com/DefinitelyTyped/DefinitelyTyped/blob/5344bfc805/types/mongoose/README.md#faq-and-common-mistakes

**Comment by human-a** [comment]
I believe this is fixed in ^14.2.1 and mongoose ^5.4.20
